src/1.generic-constraints/index.ts

```diff
@@ -4,6 +4,19 @@
 // You can remove the readonly modifier from the mapped type by using -readonly
 
 // ------- start coding here -------
+type ObjectType = {
+  [O in any]: any;
+};
+
+function mapObjectProperties<
+  S extends ObjectType,
+  M extends { [K in keyof M]: keyof S }
+>(source: S, mapping: M): { [K in keyof M]: S[M[K]] } {
```

Comment:
Return type nya salah.

Pertama, kita perlu menghilangkan readonly seperti ini:
-readonly [K in keyof M]: S[M[K]]

Kedua, kalau propertinya tidak ada di mapping, dengan cara seperti itu maka original propertynya akan terhapus.

Ketiga, kalau return typenya cukup complex, lebih baik dibuat menjadi sebuah type alias seperti ini:
```
 type Mapping<S, M extends Record<string, keyof S>> = {
  -readonly [K in keyof M]: S[M[K]];
} & Omit<S, M[keyof M]>;
```

Dengan cara ini return type nya jadi rapi karena kita cukup panggil `: Mapping<S, M>` saja.

Nah, kode intersect berikut gunanya supaya original propertynya tidak hilang saat kita tidak memasukkannya di mapping.
```
& Omit<S, M[keyof M]>
```

Contohnya:
```
const mapping2 = {
  id: 'userId',
  name: 'userName',
} as const;
```

Disini userAge nya tidak kita mapping, tapi ga usah khawatir karena kita intersect dengan `& Omit<S, M[keyof M]>`